Use probability sampling over periodic sampling

vlkale opened the pull request:
Fix #180 . 

This PR is related to old PR #181 

The current Kokkos sampler utility uses periodic sampling via a sampler skip rate. Doing this is often restrictive when sampling profiling and debugging data. For example, doing this can miss out on important data not in the periodicity of kernel invocations. The goal of this PR for the Kokkos sampler  utility is to allow user to use random sampling primarily and periodic sampling secondarily via environment variables in the form `KOKKOS_TOOLS_SAMPLER_xyz`. 

Since the solution should not allow for a combination of both periodicity and probability, the probability will always be chosen. 

For example, let us say that a user requests a a `Kokkos::parallel_for()` every 20th invocation of that `Kokkos::parallel_for()` and requests gather time spent on `Kokkos::parallel_for()` with probability 63% on each invocation of that `Kokkos::parallel_for()`. Then, the sampler will not _skip *trying* to time_ any invocations of the Kokkos::parallel_for() but it _will_ obtain a timing with probability 63% on each invocation of that `Kokkos::parallel_for()`. 

See the `common/kokkos-sampler/README.md` directory for a high-level overview - in English - of the changes.

For later: I will put in slide in the Kokkos Tools tutorial slide on sampling and filtering to explain how to use these utilities.

vlkale committed:
kp_sampler_skip: putting in probabilistic sampling

vlkale committed:
CMakeLists.txt in kokkos-sampler: fix add_library to kp_add_library

vlkale committed:
sampler's README.md: reflects sampler probability env var

vlkale committed:
kp_sampler_skip.cpp: tool_prob_num default to -1.0 not 0.1

vlkale commented:
The following two outputs show that the sampling with probability 1.0% works properly when applied to kernel timer Kokkos tool for stream benchmark in the Kokkos core benchmark folder. 

The first output is with Kokkos tools global fences being on (tool-induced fencing is enabled) and the second output is with global fencing turned off. In the second case, fencing is not invoked, as expected. Also, note that the set of Kokkos kernel invocation numbers that is sampled is different across these two different runs. The random number generator is seeded with the current time, making the invocations sampled different. The following is run on a MacOS with gcc and Kokkos 4.1. 



```
vlkale@s1088602ca stream % export KOKKOS_TOOLS_SAMPLER_VERBOSE=2; export KOKKOS_TOOLS_SAMPLER_PROB=1.0; export KOKKOS_TOOLS_GLOBALFENCES=1; export KOKKOS_TOOLS_LIBS="/Users/vlkale/Desktop/vlap/wk/code/softwareTech/ktools/ktov105/common/kokkos-sampler/kp_sampler.so;/Users/vlkale/Desktop/vlap/wk/code/softwareTech/ktools/ktov105/profiling/simple-kernel-timer/kp_kernel_timer.so"; ./stream.exe 
-------------------------------------------------------------
Kokkos STREAM Benchmark
-------------------------------------------------------------
KokkosP: Next library to call: /Users/vlkale/Desktop/vlap/wk/code/softwareTech/ktools/ktov105/profiling/simple-kernel-timer/kp_kernel_timer.so
KokkosP: Loading child library ..
KokkosP: Simple Kernel Timer Library Initialized (sequence is 1, version: 20211015)
KokkosP: Function Status:
KokkosP: begin-parallel-for:      yes
KokkosP: begin-parallel-scan:     yes
KokkosP: begin-parallel-reduce:   yes
KokkosP: end-parallel-for:        yes
KokkosP: end-parallel-scan:       yes
KokkosP: end-parallel-reduce:     yes
KokkosP: Sampling rate set to: 20
KokkosP: Sampling rate provided as input: 20
KokkosP: Sampling probability provided as input: 1.0
KokkosP: Sampling rate set to: 21
KokkosP: Sampling probability set to 1.000000
KokkosP: seeding Random Number Generator using clock for probabilistic sampling.
KokkosP: Note that both probability and skip rate are set. The Kokkos Tools Sampler utility will invoke a Kokkos Tool child event you specified (e.g., the profiler or debugger tool connector you specified in KOKKOS_TOOLS_LIBS) with only specified sampling probability applied and sampling skip rate set is ignored with no predefined periodicity for sampling used.
KokkosP: The skip rate in the sampler utility is being set to 1.
Reports fastest timing per kernel
Creating Views...
Memory Sizes:
- Array Size:    100000000
- Per Array:           800.00 MB
- Total:              2400.00 MB
Benchmark kernels will be performed for 200 iterations.
-------------------------------------------------------------
Initializing Views...
Starting benchmarking...
KokkosP: sample 12 calling child-begin function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 12 calling child-end function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 267 calling child-begin function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 267 calling child-end function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 362 calling child-begin function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 362 calling child-end function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 468 calling child-begin function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 468 calling child-end function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 503 calling child-begin function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 503 calling child-end function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 579 calling child-begin function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 579 calling child-end function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 657 calling child-begin function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 657 calling child-end function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 925 calling child-begin function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
KokkosP: sample 925 calling child-end function...
KokkosP: Sampler utility sucessfully invoked  tool-induced fence on device 0
Performing validation...
All solutions checked and verified.
-------------------------------------------------------------
Set                62629.87 MB/s
Copy               74669.24 MB/s
Scale              74154.04 MB/s
Add                83099.73 MB/s
Triad              82674.88 MB/s
-------------------------------------------------------------
KokkosP: Kernel timing written to /Users/vlkale/Desktop/vlap/wk/code/softwareTech/kokkos/benchmarks/stream/s1088602ca-41172.dat 
vlkale@s1088602ca stream % export KOKKOS_TOOLS_SAMPLER_VERBOSE=2; export KOKKOS_TOOLS_SAMPLER_PROB=1.0; export KOKKOS_TOOLS_GLOBALFENCES=0; export KOKKOS_TOOLS_LIBS="/Users/vlkale/Desktop/vlap/wk/code/softwareTech/ktools/ktov105/common/kokkos-sampler/kp_sampler.so;/Users/vlkale/Desktop/vlap/wk/code/softwareTech/ktools/ktov105/profiling/simple-kernel-timer/kp_kernel_timer.so"; ./stream.exe
-------------------------------------------------------------
Kokkos STREAM Benchmark
-------------------------------------------------------------
KokkosP: Next library to call: /Users/vlkale/Desktop/vlap/wk/code/softwareTech/ktools/ktov105/profiling/simple-kernel-timer/kp_kernel_timer.so
KokkosP: Loading child library ..
KokkosP: Simple Kernel Timer Library Initialized (sequence is 1, version: 20211015)
KokkosP: Function Status:
KokkosP: begin-parallel-for:      yes
KokkosP: begin-parallel-scan:     yes
KokkosP: begin-parallel-reduce:   yes
KokkosP: end-parallel-for:        yes
KokkosP: end-parallel-scan:       yes
KokkosP: end-parallel-reduce:     yes
KokkosP: Sampling rate set to: 20
KokkosP: Sampling rate provided as input: 20
KokkosP: Sampling probability provided as input: 1.0
KokkosP: Sampling rate set to: 21
KokkosP: Sampling probability set to 1.000000
KokkosP: seeding Random Number Generator using clock for probabilistic sampling.
KokkosP: Note that both probability and skip rate are set. The Kokkos Tools Sampler utility will invoke a Kokkos Tool child event you specified (e.g., the profiler or debugger tool connector you specified in KOKKOS_TOOLS_LIBS) with only specified sampling probability applied and sampling skip rate set is ignored with no predefined periodicity for sampling used.
KokkosP: The skip rate in the sampler utility is being set to 1.
Reports fastest timing per kernel
Creating Views...
Memory Sizes:
- Array Size:    100000000
- Per Array:           800.00 MB
- Total:              2400.00 MB
Benchmark kernels will be performed for 200 iterations.
-------------------------------------------------------------
Initializing Views...
Starting benchmarking...
KokkosP: sample 28 calling child-begin function...
KokkosP: sample 28 calling child-end function...
KokkosP: sample 296 calling child-begin function...
KokkosP: sample 296 calling child-end function...
KokkosP: sample 370 calling child-begin function...
KokkosP: sample 370 calling child-end function...
KokkosP: sample 377 calling child-begin function...
KokkosP: sample 377 calling child-end function...
KokkosP: sample 476 calling child-begin function...
KokkosP: sample 476 calling child-end function...
KokkosP: sample 503 calling child-begin function...
KokkosP: sample 503 calling child-end function...
KokkosP: sample 601 calling child-begin function...
KokkosP: sample 601 calling child-end function...
KokkosP: sample 693 calling child-begin function...
KokkosP: sample 693 calling child-end function...
KokkosP: sample 944 calling child-begin function...
KokkosP: sample 944 calling child-end function...
Performing validation...
All solutions checked and verified.
-------------------------------------------------------------
Set                62633.34 MB/s
Copy               74470.13 MB/s
Scale              74797.96 MB/s
Add                83071.21 MB/s
Triad              82997.36 MB/s
-------------------------------------------------------------
KokkosP: Kernel timing written to /Users/vlkale/Desktop/vlap/wk/code/softwareTech/kokkos/benchmarks/stream/s1088602ca-41194.dat 
vlkale@s1088602ca stream %

```

crtrott commented on common/kokkos-sampler/kp_sampler_skip.cpp:
```diff
@@ -182,6 +185,79 @@ void kokkosp_init_library(const int loadSeq, const uint64_t interfaceVer,
+        "KokkosP: seeding Random Number Generator using clock for "
+        "probabilistic sampling.\n");
+  }
+  srand(time(NULL));
```
Let have an option to set this via environment variable. `KOKKOS_TOOLS_SEED` environment variable, if not set then `srand(time(NULL));` else `srand(seed)`.

crtrott replied:
Move this above the output.

vlkale replied:
Fixed.

crtrott commented on common/kokkos-sampler/kp_sampler_skip.cpp:
```diff
@@ -221,7 +299,6 @@ void kokkosp_end_parallel_for(const uint64_t kID) {
         invoke_ktools_fence(0);
       }
       (*endForCallee)(retrievedNestedkID);
-      infokIDSample.erase(kID);
```
this line shouldn't be removed

vlkale replied:
fixed

crtrott reviewed (changes requested):
Lets add a user option to set the seed, and don't delete the erase

vlkale committed:
Update kp_sampler_skip.cpp: put erase back 

Erase got deleted (there on the develop branch) 

Took it out due to copy-paste

vlkale committed:
Update kp_sampler_skip.cpp: putting in seed 

User can input seed for RNG for probabilistic sampling

Commit:
kp_sampler: set default seed to -1, tool fence verbosity

vlkale committed:
kp_sampler_skip.cpp: applied clang format

vlkale commented:
The below is a test with the most recently committed version with KOKKOS_TOOLS_SEED set, as requested by @crtrott. Two separate runs of the program stream were done with the following environment variables set. As seen from the output of both runs, they both have the same sequence of events sampled, showing that the manual seed rather than time-generated seed is working for this case. Note that the output is truncated for easily viewing the output within this GitHub issue.

```
ViveksMacBook: stream % export KOKKOS_TOOLS_LIBS="/Users/vivek/kto-inst/libkp_kokkos_sampler.dylib;/Users/Vivek/kto-inst/libkp_kernel_logger.dylib"; export KOKKOS_TOOLS_SEED=4; export KOKKOS_TOOLS_SEED=4; export KOKKOS_TOOLS_SAMPLER_VERBOSE=2; export KOKKOS_TOOLS_SAMPLER_PROBABILITY=50.0; ./stream.exe;
-------------------------------------------------------------
Kokkos STREAM Benchmark
-------------------------------------------------------------
KokkosP: Next library to call: /Users/vivek/kto-inst/libkp_kernel_logger.dylib
KokkosP: Loading child library ..
KokkosP: Kernel Logger Library Initialized (sequence is 1, version: 20211015)
KokkosP: Function Status:
KokkosP: begin-parallel-for:      yes
KokkosP: begin-parallel-scan:     yes
KokkosP: begin-parallel-reduce:   yes
KokkosP: end-parallel-for:        yes
KokkosP: end-parallel-scan:       no
KokkosP: end-parallel-reduce:     yes
KokkosP: Sampling rate set to: 1
KokkosP: Sampling rate provided as input: 1
KokkosP: Sampling probability provided as input: 50.0
KokkosP: Sampling rate set to: 2
KokkosP: Sampling probability set to 50.000000
KokkosP: Seeding random number generator using seed 4 for probabilistic sampling.
KokkosP: Note that both probability and skip rate are set. The Kokkos Tools Sampler utility will invoke a Kokkos Tool child event you specified (e.g., the profiler or debugger tool connector you specified in KOKKOS_TOOLS_LIBS) with only specified sampling probability applied and sampling skip rate set is ignored with no predefined periodicity for sampling used.
KokkosP: The skip rate in the sampler utility is being set to 1.
Reports fastest timing per kernel
Creating Views...
Memory Sizes:
- Array Size:    100000000
- Per Array:           800.00 MB
- Total:              2400.00 MB
Benchmark kernels will be performed for 20 iterations.
-------------------------------------------------------------
KokkosP: sample 1 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 0
KokkosP:     Kokkos::View::initialization [a] via memset
KokkosP: sample 1 finished with child-begin function.
KokkosP: sample 1 calling child-end function...
KokkosP: Execution of kernel 0 is completed.
KokkosP: sample 3 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 1
KokkosP:     Kokkos::View::initialization [c] via memset
KokkosP: sample 3 finished with child-begin function.
KokkosP: sample 3 calling child-end function...
KokkosP: Execution of kernel 1 is completed.
Initializing Views...
Starting benchmarking...
KokkosP: sample 5 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 2
KokkosP:     set
KokkosP: sample 5 finished with child-begin function.
KokkosP: sample 5 calling child-end function...
KokkosP: Execution of kernel 2 is completed.

...

KokkosP: sample 96 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 47
KokkosP:     copy
KokkosP: sample 96 finished with child-begin function.
KokkosP: sample 96 calling child-end function...
KokkosP: Execution of kernel 47 is completed.
KokkosP: sample 98 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 48
KokkosP:     add
KokkosP: sample 98 finished with child-begin function.
KokkosP: sample 98 calling child-end function...
KokkosP: Execution of kernel 48 is completed.
KokkosP: sample 99 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 49
KokkosP:     triad
KokkosP: sample 99 finished with child-begin function.
KokkosP: sample 99 calling child-end function...
KokkosP: Execution of kernel 49 is completed.
KokkosP: sample 101 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 50
KokkosP:     copy
KokkosP: sample 101 finished with child-begin function.
KokkosP: sample 101 calling child-end function...
KokkosP: Execution of kernel 50 is completed.
KokkosP: sample 103 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 51
KokkosP:     add
KokkosP: sample 103 finished with child-begin function.
KokkosP: sample 103 calling child-end function...
KokkosP: Execution of kernel 51 is completed.
Performing validation...
All solutions checked and verified.
-------------------------------------------------------------
Set                12593.03 MB/s
Copy               16127.57 MB/s
Scale              15996.91 MB/s
Add                16634.53 MB/s
Triad              17182.65 MB/s
-------------------------------------------------------------
KokkosP: Kokkos library finalization called.
ViveksMacBook stream % ./stream.exe
-------------------------------------------------------------
Kokkos STREAM Benchmark
-------------------------------------------------------------
KokkosP: Next library to call: /Users/Vivek/kto-inst/libkp_kernel_logger.dylib
KokkosP: Loading child library ..
KokkosP: Kernel Logger Library Initialized (sequence is 1, version: 20211015)
KokkosP: Function Status:
KokkosP: begin-parallel-for:      yes
KokkosP: begin-parallel-scan:     yes
KokkosP: begin-parallel-reduce:   yes
KokkosP: end-parallel-for:        yes
KokkosP: end-parallel-scan:       no
KokkosP: end-parallel-reduce:     yes
KokkosP: Sampling rate set to: 1
KokkosP: Sampling rate provided as input: 1
KokkosP: Sampling probability provided as input: 50.0
KokkosP: Sampling rate set to: 2
KokkosP: Sampling probability set to 50.000000
KokkosP: Seeding random number generator using seed 4 for probabilistic sampling.
KokkosP: Note that both probability and skip rate are set. The Kokkos Tools Sampler utility will invoke a Kokkos Tool child event you specified (e.g., the profiler or debugger tool connector you specified in KOKKOS_TOOLS_LIBS) with only specified sampling probability applied and sampling skip rate set is ignored with no predefined periodicity for sampling used.
KokkosP: The skip rate in the sampler utility is being set to 1.
Reports fastest timing per kernel
Creating Views...
Memory Sizes:
- Array Size:    100000000
- Per Array:           800.00 MB
- Total:              2400.00 MB
Benchmark kernels will be performed for 20 iterations.
-------------------------------------------------------------
KokkosP: sample 1 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 0
KokkosP:     Kokkos::View::initialization [a] via memset
KokkosP: sample 1 finished with child-begin function.
KokkosP: sample 1 calling child-end function...
KokkosP: Execution of kernel 0 is completed.
KokkosP: sample 3 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 1
KokkosP:     Kokkos::View::initialization [c] via memset
KokkosP: sample 3 finished with child-begin function.
KokkosP: sample 3 calling child-end function...
KokkosP: Execution of kernel 1 is completed.
Initializing Views...
Starting benchmarking...
KokkosP: sample 5 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 2
KokkosP:     set
KokkosP: sample 5 finished with child-begin function.
KokkosP: sample 5 calling child-end function...
KokkosP: Execution of kernel 2 is completed.

....

KokkosP: sample 96 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 47
KokkosP:     copy
KokkosP: sample 96 finished with child-begin function.
KokkosP: sample 96 calling child-end function...
KokkosP: Execution of kernel 47 is completed.
KokkosP: sample 98 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 48
KokkosP:     add
KokkosP: sample 98 finished with child-begin function.
KokkosP: sample 98 calling child-end function...
KokkosP: Execution of kernel 48 is completed.
KokkosP: sample 99 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 49
KokkosP:     triad
KokkosP: sample 99 finished with child-begin function.
KokkosP: sample 99 calling child-end function...
KokkosP: Execution of kernel 49 is completed.
KokkosP: sample 101 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 50
KokkosP:     copy
KokkosP: sample 101 finished with child-begin function.
KokkosP: sample 101 calling child-end function...
KokkosP: Execution of kernel 50 is completed.
KokkosP: sample 103 calling child-begin function...
KokkosP: Executing parallel-for kernel on device 100663297 with unique execution identifier 51
KokkosP:     add
KokkosP: sample 103 finished with child-begin function.
KokkosP: sample 103 calling child-end function...
KokkosP: Execution of kernel 51 is completed.
Performing validation...
All solutions checked and verified.
-------------------------------------------------------------
Set                12432.48 MB/s
Copy               15896.57 MB/s
Scale              15927.12 MB/s
Add                17134.47 MB/s
Triad              17060.44 MB/s
-------------------------------------------------------------
KokkosP: Kokkos library finalization called.

```

vlkale committed:
Sampler: reduce text in prints; revise README

vlkale committed:
fix readme indent

vlkale commented:
> Lets add a user option to set the seed, and don't delete the erase

Done.

crtrott commented on common/kokkos-sampler/kp_sampler_skip.cpp:
```diff
@@ -78,6 +88,8 @@ void kokkosp_init_library(const int loadSeq, const uint64_t interfaceVer,
                           const uint32_t devInfoCount, void* deviceInfo) {
   const char* tool_verbose_str   = getenv("KOKKOS_TOOLS_SAMPLER_VERBOSE");
   const char* tool_globFence_str = getenv("KOKKOS_TOOLS_GLOBALFENCES");
+  const char* tool_seed_str      = getenv("KOKKOS_TOOLS_SEED");
```
Lets maybe name it KOKKOS_TOOLS_RANDOM_SEED

crtrott commented on common/kokkos-sampler/kp_sampler_skip.cpp:
```diff
@@ -88,6 +100,11 @@ void kokkosp_init_library(const int loadSeq, const uint64_t interfaceVer,
+  if (NULL != tool_seed_str) {
+    tool_seed = atoi(tool_seed_str);
+  } else {
+    tool_seed = 1;
```
Why? don't you check below for tool_seed < 0 to see if you want to set it via time?

vlkale replied:
Fixed. Thanks.

vlkale committed:
kp_sampler: fix name of seed environment varable and else condition for seed for user input + README

vlkale committed:
Update kp_sampler_skip.cpp: fix msg Kokkos_tools_libs

vlkale committed:
Merge branch 'kokkos:develop' into use-probability-sampling

vlkale committed:
intitialized pointer

vlkale committed:
kp_sampler_skip.cpp: revert fix of initialize pointer

vlkale committed:
Update kp_sampler_skip.cpp: fix func pointer - tools

vlkale committed:
kp_sampler_skip.cpp: provide_tpi parameter changed

vlkale committed:
kp_core.hpp: provide_tpi takes struct

vlkale committed:
Update kp_sampler_skip.cpp: fix paren in invoke_ktools_fence

vlkale committed:
kp_sampler_skip.cpp: applied clang-format-8

vlkale committed:
kp_core.hpp: applied clang-format-8

vlkale commented:
For additional example run using the Kokkos Tools Sampler with randomized sampling, below is the build and subsequent output with sampler utility applied to the Kernel logger tool connector library run with the stream.cuda executable on Perlmutter for 200 outer iterations, i.e.,  200 timesteps. The sampling is done at 1.2% probability, i.e., for an invocation of any Kokkos kernel in the program, there is a 1.2% chance that the kernel will be printed/logged on the screen to a user. 

The below output is reproducible on Perlmutter through using `module load PrgEnv-gnu`, building with the Kokkos develop branch and using the Kokkos Serial+CUDA backend (build line from this Kokkos Tools PR is shown below before the output),  setting the `KOKKOS_TOOLS_RANDOM_SEED` variable to 2 (if you set this variable to another number you will get another set of invocations that are in the sampled set). 


# Build  
```

vkale3@perlmutter:login29:~/kto-dev-vlk/nvtxbld> ccmake .. -DCMAKE_INSTALL_PREFIX="/global/u2/v/vkale3/ktins20240222-2" -DKokkos_COMPILE_LAUNCHER=/global/u2/v/vkale3/kks/bin/kokkos_launch_compiler -DKokkos_DIR=/global/u2/v/vkale3/kks/kbuild-cuda -DKokkos_NVCC_WRAPPER="/global/u2/v/vkale3/kks/bin/nvcc_wrapper" 

```

# Output
```
vkale3@perlmutter:login13:~/kks/benchmarks/stream> export KOKKOS_TOOLS_LIBS="/global/u2/v/vkale3/ktins20240222-2/lib64/libkp_kokkos_sampler.so;/global/u2/v/vkale3/ktins20240222-2/lib64/libkp_kernel_logger.so"; export KOKKOS_TOOLS_SAMPLER_VERBOSE=2; export KOKKOS_TOOLS_RANDOM_SEED=2; export KOKKOS_TOOLS_GLOBALFENCES=1; export KOKKOS_TOOLS_SAMPLER_PROB=1.2; /global/homes/v/vkale3/kks/benchmarks/stream/stream.cuda  
-------------------------------------------------------------
Kokkos STREAM Benchmark
-------------------------------------------------------------
KokkosP: Next library to call: /global/u2/v/vkale3/ktins20240222-2/lib64/libkp_kernel_logger.so
KokkosP: Loading child library ..
KokkosP: Kernel Logger Library Initialized (sequence is 1, version: 20211015)
KokkosP: Function Status:
KokkosP: begin-parallel-for:      yes
KokkosP: begin-parallel-scan:     yes
KokkosP: begin-parallel-reduce:   yes
KokkosP: end-parallel-for:        yes
KokkosP: end-parallel-scan:       no
KokkosP: end-parallel-reduce:     yes
KokkosP: Sampling rate set to: 20
KokkosP: Sampling skip rate provided as input is: 20
KokkosP: Sampling probability provided as input is: 1.2
KokkosP: Sampling skip rate is set to: 21
KokkosP: Sampling probability is set to 1.200000
KokkosP: Seeding random number generator using seed 2 for random sampling.
KokkosP: You set both the probability and skip rate for the sampler. Only random sampling will be done, using the probabability you set; The skip rate you set will be ignored.
KokkosP: Note: The skip rate will be set to 1. Sampling will not be based  on a pre-defined periodicity.
Kokkos: Kokkos_Profiling.cpp:initialize: actions.fence = 0x10f6390 	 fenceFnPtr = (nil) 
Kokkos: Kokkos_Profiling.cpp:initialize: actions 0x7fff6d1179d0 	 fence address 0x10f6390
Kokkos: Kokkos_Profiling.cpp:initialize: tool_invoked_fence address 0x41a8f0 	 fenceFnPtr 0x41a8f0 
Kokkos: Kokkos_Profiling.cpp:initialize: after fence init actions 0x7fff6d1179d0 	 fence address 0x41a8f0
Reports fastest timing per kernel
Creating Views...
Memory Sizes:
- Array Size:    1000
- Per Array:             0.01 MB
- Total:                 0.02 MB
Benchmark kernels will be performed for 200 iterations.
-------------------------------------------------------------
Initializing Views...
Starting benchmarking...
KokkosP: sample 220 calling child-begin function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Executing parallel-for kernel on device 33554433 with unique execution identifier 0
KokkosP:     scale
KokkosP: sample 220 finished with child-begin function.
KokkosP: sample 220 calling child-end function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Execution of kernel 0 is completed.
KokkosP: sample 246 calling child-begin function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Executing parallel-for kernel on device 33554433 with unique execution identifier 1
KokkosP:     add
KokkosP: sample 246 finished with child-begin function.
KokkosP: sample 246 calling child-end function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Execution of kernel 1 is completed.
KokkosP: sample 304 calling child-begin function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Executing parallel-for kernel on device 33554433 with unique execution identifier 2
KokkosP:     copy
KokkosP: sample 304 finished with child-begin function.
KokkosP: sample 304 calling child-end function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Execution of kernel 2 is completed.
KokkosP: sample 403 calling child-begin function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Executing parallel-for kernel on device 33554433 with unique execution identifier 3
KokkosP:     set
KokkosP: sample 403 finished with child-begin function.
KokkosP: sample 403 calling child-end function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Execution of kernel 3 is completed.
KokkosP: sample 528 calling child-begin function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Executing parallel-for kernel on device 33554433 with unique execution identifier 4
KokkosP:     set
KokkosP: sample 528 finished with child-begin function.
KokkosP: sample 528 calling child-end function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Execution of kernel 4 is completed.
KokkosP: sample 625 calling child-begin function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Executing parallel-for kernel on device 33554433 with unique execution identifier 5
KokkosP:     scale
KokkosP: sample 625 finished with child-begin function.
KokkosP: sample 625 calling child-end function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Execution of kernel 5 is completed.
KokkosP: sample 642 calling child-begin function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Executing parallel-for kernel on device 33554433 with unique execution identifier 6
KokkosP:     triad
KokkosP: sample 642 finished with child-begin function.
KokkosP: sample 642 calling child-end function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Execution of kernel 6 is completed.
KokkosP: sample 737 calling child-begin function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Executing parallel-for kernel on device 33554433 with unique execution identifier 7
KokkosP:     triad
KokkosP: sample 737 finished with child-begin function.
KokkosP: sample 737 calling child-end function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Execution of kernel 7 is completed.
KokkosP: sample 804 calling child-begin function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Executing parallel-for kernel on device 33554433 with unique execution identifier 8
KokkosP:     copy
KokkosP: sample 804 finished with child-begin function.
KokkosP: sample 804 calling child-end function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Execution of kernel 8 is completed.
KokkosP: sample 849 calling child-begin function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Executing parallel-for kernel on device 33554433 with unique execution identifier 9
KokkosP:     copy
KokkosP: sample 849 finished with child-begin function.
KokkosP: sample 849 calling child-end function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Execution of kernel 9 is completed.
KokkosP: sample 927 calling child-begin function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Executing parallel-for kernel on device 33554433 with unique execution identifier 10
KokkosP:     triad
KokkosP: sample 927 finished with child-begin function.
KokkosP: sample 927 calling child-end function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Execution of kernel 10 is completed.
KokkosP: sample 957 calling child-begin function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Executing parallel-for kernel on device 33554433 with unique execution identifier 11
KokkosP:     triad
KokkosP: sample 957 finished with child-begin function.
KokkosP: sample 957 calling child-end function...
KokkosP: Sampler attempting to invoke tool-induced fence on device 0.
KokkosP: Sampler sucessfully invoked tool-induced fence on device 0
KokkosP: Execution of kernel 11 is completed.
Performing validation...
All solutions checked and verified.
-------------------------------------------------------------
Set                  956.25 MB/s
Copy                1892.15 MB/s
Scale               1852.71 MB/s
Add                 2750.40 MB/s
Triad               2808.33 MB/s
-------------------------------------------------------------
KokkosP: Kokkos library finalization called.

```

vlkale commented:
> Lets add a user option to set the seed, and don't delete the erase
This is put in via KOKKOS_TOOLS_RANDOM_SEED

vlkale committed:
Merge branch 'kokkos:develop' into use-probability-sampling

vlkale committed:
Create test_randomized_sampling.cpp

vlkale committed:
Create CMakeLists.txt

vlkale committed:
Update test_randomized_sampling.cpp

crtrott commented on tests/sampler/test_randomized_sampling.cpp:
```diff
@@ -0,0 +1,69 @@
+
+#include "Kokkos_Core.hpp"
+
+#define NUMITERS 1000
```
no reason to do this.

crtrott commented on tests/sampler/test_randomized_sampling.cpp:
```diff
@@ -0,0 +1,69 @@
+  KOKKOS_FUNCTION void operator()(const int) const {}
+
+  template <typename TagType>
+  KOKKOS_FUNCTION void operator()(const TagType, const int) const {}
```
You are not using tags

vlkale committed:
test_randomized_sampling.cpp: fix matcher text for 4 samples

vlkale committed:
Update CMakeLists.txt: add sampler subdirectory

vlkale committed:
Update CMakeLists.txt: fix to be kp_kokkos_sampler

vlkale committed:
Update test_randomized_sampling.cpp: fix to test with RE2

vlkale committed:
Update test_randomized_sampling.cpp: vector include

vlkale committed:
Update test_randomized_sampling.cpp: curly fix

vlkale committed:
Update test_randomized_sampling.cpp: RE2 quotation fix

vlkale committed:
CMakeLists.txt: put in test for kernel logger

vlkale committed:
test_randomized_sampling.cpp: sample number to be 100

vlkale committed:
test_randomized_sampling.cpp: fix comment on test

vlkale committed:
test_randomized_sampling.cpp: env var default

vlkale committed:
Update CMakeLists.txt: fix sampler

vlkale committed:
Update CMakeLists.txt: put add'l env variables

vlkale committed:
test_randomized_sampling.cpp: apply clang format

vlkale committed:
CMakeLists.txt: put in sampler verbose

vlkale committed:
test_randomized_sampling: apply clang format

vlkale committed:
test_randomized_sampling.cpp: fixing setting environment variable in Ctest

vlkale committed:
test_randomized_sampling.cpp: apply clang-format

vlkale committed:
fix test randomized sampling

vlkale committed:
CMakeLists.txt: fix enviornment var for KOKKOS_TOOLS_LIBS2

vlkale committed:
CMakeLists.txt: add global fences environment variable

vlkale committed:
CMakeLists.txt: fix nesting of if's in ctest

vlkale committed:
Update CMakeLists.txt: kaeat_args

vlkale committed:
Update CMakeLists.txt: kp_kernel_logger fix

vlkale committed:
Update CMakeLists.txt: fix target file 2

vlkale committed:
add cmake

vlkale committed:
Merge branch 'use-probability-sampling' of https://github.com/vlkale/kokkos-tools into use-probability-sampling

vlkale committed:
random sampling

vlkale committed:
test_randomized_sampling.cpp: fix matcher

vlkale committed:
Update CMakeLists.txt: fix sampler skip variable

vlkale committed:
Update CMakeLists.txt: env var

vlkale committed:
Update kp_sampler_skip.cpp: remove else kernelSampleSkip

vlkale committed:
Update kp_sampler_skip.cpp: period in sampler fence printf

vlkale committed:
kp_sampler_skip.cpp: removed tool verbose str print

vlkale committed:
test_randomized_sampling.cpp: take out flushes

masterleinad committed:
Convert kp_kernel_logger to std::cout

masterleinad committed:
Working

masterleinad committed:
Almost working

masterleinad committed:
Still working

masterleinad committed:
Use cout in sampler

masterleinad committed:
Revert changes to kp_kernel_logger.cpp

masterleinad committed:
More indentation

vlkale committed:
kp_sampler_skip.cpp: apply clang format

vlkale committed:
test_randomized_sampling.cpp: apply clang format

vlkale committed:
Merge pull request #9 from masterleinad/use-probability-sampling

Fix https://github.com/kokkos/kokkos-tools/pull/213 - Merge printf conversion to cout in sampler for the sake of Gtest.  We may want to provide an option of std::cout or printf for the user in the future. I am keeping this as is for now.

vlkale committed:
kp_sampler_skip.cpp: fix reduce callee null check

vlkale committed:
Merge branch 'kokkos:develop' into use-probability-sampling

vlkale committed:
kp_kernel_logger.cpp: scan typo fix

vlkale committed:
Rename test_randomized_sampling.cpp to test_parfor.cpp

vlkale committed:
Update CMakeLists.txt: fix sampling

vlkale committed:
Prob smplr (#16)

* CMakeLists.txt: add_library to kp_add_library

Sampler's CMakeLists.txt: add_library to kp_add_library

* kp_core.hpp: remove pointer from ptpi

* kp_sampler_skip.cpp: pass by value kokkosp_p_t_p_i

kp_sampler_skip.cpp: last parameter should be passed by value rather than pointer in kokkosp_p_t_p_i

* kp_sampler_skip.cpp: remove dereference assignment for ptpi callback

* kp_core.hpp: apply clang-format

* Create test_sampler.cpp

* CMakeLists.txt: set up test for sampler

* test_sampler.cpp: put in code for sampler test

* test_sampler.cpp: typo in comment (2 invocation --> 2 invocations)

* CMakeLists.txt: support sampler test

* kp_sampler_skip.cpp: change printf to std::out for ctests

* kp_sampler_skip.cpp: include iostream for std::cout

* Update kp_sampler_skip.cpp: \n instead of std::endl

\n is faster for performance

* kp_sampler_skip.cpp: apply clang format

* kp_sampler_skip.cpp: fix for std::out of tool-invoked fence verbose debug print

* kp_sampler_skip.cpp: apply clang format

* Rename test_sampler.cpp to test_parfor.cpp

* Update CMakeLists.txt: reduce and scan sampling tests

* Create test_parreduce.cpp

* Create test_parscan.cpp

* Update test_parreduce.cpp: parallel_reduce function 

Test based on example shown at: https://kokkos.org/kokkos-core-wiki/API/core/parallel-dispatch/parallel_reduce.html

* test_parscan.cpp: put in test for sampling parallel_scan

Parallel scan sampling test based on example here: 

https://kokkos.org/kokkos-core-wiki/API/core/parallel-dispatch/parallel_scan.html

* Update test_parreduce.cpp: fix x[i] to x(i) 

View  index access of x does not use []

* Update test_parreduce.cpp: Kokkos:: for View

* test_parreduce.cpp: reduce lambda

* Update test_parscan.cpp: operator for cuda/hip build

* test_parscan.cpp: support scan test function

* test_parreduce.cpp: reduction operator second argument 

second argument

* test_parscan.cpp: fix scan test operator

* Update test_parscan.cpp: fix scan test to have to Views

* test_parscan.cpp: policy to size

* Update kp_kernel_logger.cpp: fix typo for scan callback

The typo was kokkospk_end_parallel_scan. This causes the third test for the sampler to fail.

* test_parscan.cpp: apply clang format

* test_parreduce.cpp: apply clang format

* CMakeLists.txt: add_library -> kp_add_library

* kp_sampler_skip.cpp: put in prob samplr

* CMakeLists.txt: sampling prob test

* Create test_parfor_prob.cpp

* Create test_parreduce_prob.cpp

* Create test_parscan_prob.cpp

* test_parfor_prob.cpp: edit matchers for prob sampling

* test_parreduce_prob.cpp: fix matchers

* Update test_parscan_prob.cpp: fix parscan test matchers

* Update test_parscan_prob.cpp: fix test comments

* test_parfor_prob.cpp: fix comments

* Update test_parreduce_prob.cpp: fix comments

* test_parfor.cpp: update matcher

* Update test_parscan.cpp: fix matchers

* test_parreduce.cpp: matchers fix

* Update kp_sampler_skip.cpp: add cout for end_parallel_xxx

* Update test_parfor.cpp: put in finished with end in matchers

* Update test_parscan.cpp: update matchers / comments removal

* test_parreduce.cpp: fix matchers

* Update kp_sampler_skip.cpp: fix elipses

* test_parscan.cpp: apply clang format

* CMakeLists.txt: skip rate is 0

* test_parscan.cpp: apply clang format

* test_parfor_prob.cpp: apply clang format

* Update README.md: add sampler entry

* test_parreduce_prob.cpp: apply clang format

* kp_sampler_skip.cpp: apply clang format

* README.md: sampler README for probability

* test_parreduce.cpp: fix int ref in operator

* test_parscan.cpp: int ref in operator

* test_parscan.cpp: long int in signature

* test_parreduce.cpp: long int in second argument to operator

* Update test_parreduce.cpp: change sum to int type

* Update test_parreduce.cpp: declare sum as long int

* Update kp_kernel_logger.cpp: kokkospk_end_parallel_scan --> kokkosp_end_parallel_scan

* Update README.md

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>

* Update test_parfor.cpp

* test_parfor.cpp: put in fence test

* Update test_parfor.cpp: put in checks for number of calls and Null Ptr fence

* test_parfor.cpp: apply clang format

* fix test par for

* test_parfor.cpp: apply clang format

* kp_sampler_skip.cpp: fix sampler std::cout prints for test

* test_parfor.cpp: apply clang format

* test_parfor.cpp: fixing matcher string for number contains

* Update test_parfor.cpp: not substr

* test_parfor.cpp: apply clang format

* kp_sampler_skip.cpp: apply clang format

* test_parfor.cpp: apply clang format

* test_parfor.cpp: apply clang format

* Update test_parfor.cpp

* test_parfor.cpp: apply clang format

* test_parfor.cpp: Times function

* delete file accidentally put in tests directory

* Update test_parfor.cpp: remove AtMost

* Update test_parfor.cpp: remove using AtMost

* kp_sampler_skip.cpp: put back in

* Update test_parreduce.cpp

* test_parfor.cpp: remove count for number of times

* Update test_parscan.cpp

* Update test_parfor.cpp: removing Contains

* kp_sampler_skip.cpp: apply clang format

* test_parfor.cpp: apply clang format

* test_parreduce.cpp: apply clang format

* test_parscan.cpp: apply clang format

* Update test_parfor.cpp: putting in times

* test_parfor.cpp: apply clang format

* Update test_parreduce.cpp: put in times

* test_parscan.cpp: put in times test

* test_parscan.cpp: apply clang format

* test_parreduce.cpp: apply clang format

* test_parscan.cpp: apply clang format

* test_parscan.cpp: apply clang format

* test_parreduce.cpp: apply clang format

* test_parfor.cpp: apply clang format

* test_parfor.cpp: apply clang format

* test_parreduce.cpp: apply clang format

* test_parscan.cpp: apply clang format

* kp_sampler_skip.cpp: std::cout to std::cerr

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>

* test_parscan.cpp: fix declaration for variable result

Co-authored-by: Damien L-G <dalg24+github@gmail.com>

* test_parfor.cpp: revisions of tests PR review

* test_parreduce.cpp: fix with new tests

* test_parscan.cpp: fix sampler scan

* test_parreduce.cpp: fix with correct samples and comments

* CMakeLists.txt: fix sampler skip rate

* test_parfor.cpp: add string header

* test_parreduce.cpp: add string header

* test_parscan.cpp: add string header

* Update test_parfor.cpp: fix output

* test_parreduce.cpp: fix test

* test_parscan.cpp: fix occurrences variable

* test_parfor.cpp: apply clang format

* test_parreduce.cpp: apply clang format

* test_parscan.cpp: apply clang format

* test_parreduce.cpp: apply clang format

* test_parfor.cpp: fix sampler number from 11 to 12

* test_parreduce.cpp: sample number from 11 to 12

* test_parscan.cpp: change sample from 11 to 12

* kp_sampler_skip.cpp: revert cerr to cout

* CMakeLists.txt: make global fences 0

* CMakeLists.txt: remove setting of global fence environment variable

* Put in tests for sampler  (#248)

* CMakeLists.txt: tests for prob sampling

Without fence and with fence

* parreduce.hpp: file for parallel reduce

* Update parreduce.hpp: putting once pragma for header

* parfor.hpp: header file for parfor

* parfor.hpp: put in code for Kokkos parallel for test

* Create parscan.hpp: parscan code file

* parscan.hpp: put in code for scan

* test_parreduce.cpp: deleting parreduce header

* test_parscan.cpp: using header for parscan

* test_parfor.cpp: parfor header

* test_parfor_prob.cpp: header

* test_parreduce_prob.cpp: putting in parreduce hpp

* test_parscan_prob.cpp: fix includes

* test_parreduce_prob.cpp: HasSubStr to HasSubstr

* new MatchersProb based on random seed two from CI build test OpenMP

* matchersProb.hpp: update

* test_par*_prob.cpp: apply clang format

* test_parfor.cpp: apply clang format

* matchersProb.hpp: fix to matchers - apply clang format

* matchersProb.hpp: fix to matchers semicolon apply clang format

* matchersProb.hpp: fix to matchers semicolon apply clang format

---------

Co-authored-by: Christian Trott <crtrott@sandia.gov>
Co-authored-by: Damien L-G <dalg24@gmail.com>
Co-authored-by: Daniel Arndt <arndtd@ornl.gov>
Co-authored-by: Damien L-G <dalg24+github@gmail.com>

vlkale committed:
Merge branch 'develop' into use-probability-sampling

vlkale committed:
test_par*.cpp: apply clang format

vlkale committed:
test_parreduce.cpp: change test tag; apply clang format

vlkale commented:
This PR now has tests for the probability sampling and is rebased with develop.

vlkale committed:
Merge branch 'kokkos:develop' into use-probability-sampling

vlkale committed:
Delete matchersProb.hpp

vlkale committed:
kp_sampler_skip.cpp: make function pointer correct

vlkale committed:
kp_sampler_skip.cpp: invoke kokkos tools fence

vlkale committed:
kp_sampler_skip.cpp: Remove small changes not relating to this PR

Fixing inconsistencies of diff with develop that shouldn't be there. These changes do not belong to this PR. They changes comprise primarily of minor changes with printed output. These unintended changes that I have taken out don't change any core logic or control flow in the program.

vlkale committed:
kp_sampler_skip.cpp: apply clang format

vlkale committed:
README.md: eliminate unintended changes in text to make diff more readable/auditable

vlkale committed:
CMakeLists.txt for test tools: comment in sampler prob environment variable should refer to that variable and not sampler skip

vlkale committed:
kp_sampler_skip.cpp: apply clang format

crtrott commented on common/kokkos-sampler/kp_sampler_skip.cpp:
```diff
@@ -1,350 +1,420 @@
+                 "sampling were set...\n";
+  }
+  tool_prob_num    = 10.0;
+  kernelSampleSkip = 1;
```
You hard code both tool_prob_num and kernelSampleSkip ... the user settings are ignored.

crtrott replied:
Put the entire thing into a `if (tool_prob_num == -1.0)`

vlkale replied:
Fixed. 

I think the key suggestion that handles the issue raised is handled by the single one you suggested in the end of this review. 

We might want to catch the case when tool_prob_num is negative but not -1.0 - I think that should happen only if there is some non-deterministic bit flip like event that causes `tool_prob_num` to be negative. 

Note I have now also put comments for each possible user input and how it is handled.

crtrott commented on common/kokkos-sampler/kp_sampler_skip.cpp:
```diff
@@ -1,350 +1,420 @@
+        << " percent. The skip rate for sampler will not be used.\n";
+  }
+
+  kernelSampleSkip = 1;
```
And again setting sampleskip ...

vlkale replied:
Deleted this. 

This is a redundant instruction in the program. 

For later: we may want to consider what to do if, e.g., a cosmic ray flips a bit in the environment variable (I think we aren't trying to build resilience in this sampler).

crtrott commented on common/kokkos-sampler/kp_sampler_skip.cpp:
```diff
@@ -1,350 +1,420 @@
+
+  const char* tool_sample = getenv("KOKKOS_TOOLS_SAMPLER_SKIP");
+  if (NULL != tool_sample) {
+    tool_prob_num    = 100.0;
```
this means SamplerSkip is prioritized over random setting, because it overwrites the previous setting

vlkale replied:
Yes, this is fixed now.

crtrott commented on common/kokkos-sampler/kp_sampler_skip.cpp:
```diff
@@ -1,350 +1,420 @@
+                << tool_prob_num << "\n";
+    }
+    kernelSampleSkip = 1;
+    return;
```
ah I see return, but none of the other stuff has returns ...

vlkale replied:
@crtrott  Please see updated version. I think your fix suggested below takes care of this.

crtrott commented on common/kokkos-sampler/kp_sampler_skip.cpp:
```diff
@@ -1,350 +1,420 @@
+  }
+
+  const char* tool_sample = getenv("KOKKOS_TOOLS_SAMPLER_SKIP");
+  if (NULL != tool_sample) {
```
```suggestion
  if (NULL != tool_sample && tool_prob_num == -1.0) {
```

vlkale committed:
kp_sampler_skip.cpp: make probability sampling the priority (as before)

Fix from a previous suggestion from contributed commit  - it may have been misunderstood that probability sampling is the priority here (I should have caught this).

Co-authored-by: Christian Trott <crtrott@sandia.gov>

vlkale committed:
kp_sampler_skip.cpp: fix conditional as per review request 

I put comments to make clear the desired behavior. Note that the conditional shouldn't be needed.

vlkale committed:
kp_sampler_skip.cpp: putting back in conditional as requested

vlkale committed:
kp_sampler_skip.cpp: apply clang-format

vlkale committed:
kp_sampler_skip.cpp: take out returns from input processing

crtrott committed:
Merge branch 'develop' into use-probability-sampling

crtrott committed:
Fix clang-format
